File: frontend/src/api/examProfile.ts
import { ApiError, api } from "./client";

export type EnglishTrack = "english_1" | "english_2";
export type MathTrack = "math_1" | "math_2" | "none";
export type SubjectCode = "english" | "math" | "politics";
export type CetStatus = "not_taken" | "cet4" | "cet6" | null;
export type MathMasteryLevel = "zero" | "basic" | "good" | "strong" | null;

export type ExamMajorCategory = {
  code: string;
  name: string;
  sort_order: number;
};

export type ExamMajor = {
  code: string;
  category_code: string;
  name: string;
  default_english_track: EnglishTrack;
  default_math_track: MathTrack;
  default_subject_codes: SubjectCode[];
  notes?: string | null;
};

export type StudentExamProfile = {
  major_category_code: string;
  major_code: string;
  english_track: EnglishTrack | null;
  math_track: MathTrack | null;
  subject_codes: SubjectCode[];
  cet_status: CetStatus;
  cet_score: number | null;
  math_mastery_level: MathMasteryLevel;
  profile_completed_at: string | null;
  created_at: string;
  updated_at: string;
};

export type UpsertStudentExamProfileInput = {
  major_category_code: string;
  major_code: string;
  english_track: EnglishTrack | null;
  math_track: MathTrack | null;
  subject_codes: SubjectCode[];
  cet_status: CetStatus;
  cet_score: number | null;
  math_mastery_level: MathMasteryLevel;
};

export function listExamMajorCategories() {
  return api<ExamMajorCategory[]>("/exam-majors/categories");
}

export function listExamMajorsByCategory(categoryCode: string) {
  return api<ExamMajor[]>(`/exam-majors?category=${encodeURIComponent(categoryCode)}`);
}

async function getExamProfileByPath(path: string) {
  try {
    return await api<StudentExamProfile>(path);
  } catch (error) {
    if (error instanceof ApiError && error.status === 404) return null;
    throw error;
  }
}

export function getAdminStudentExamProfile(studentId: string) {
  return getExamProfileByPath(`/admin/students/${studentId}/exam-profile`);
}

export function saveAdminStudentExamProfile(
  studentId: string,
  body: UpsertStudentExamProfileInput,
) {
  return api<StudentExamProfile>(`/admin/students/${studentId}/exam-profile`, {
    method: "PUT",
    body: JSON.stringify(body),
  });
}

export function confirmAdminStudentExamProfile(studentId: string) {
  return api<StudentExamProfile>(`/admin/students/${studentId}/exam-profile/confirm`, {
    method: "POST",
  });
}

export function getStaffStudentExamProfile(studentId: string) {
  return getExamProfileByPath(`/staff/students/${studentId}/exam-profile`);
}

export function saveStaffStudentExamProfile(
  studentId: string,
  body: UpsertStudentExamProfileInput,
) {
  return api<StudentExamProfile>(`/staff/students/${studentId}/exam-profile`, {
    method: "PUT",
    body: JSON.stringify(body),
  });
}

export function confirmStaffStudentExamProfile(studentId: string) {
  return api<StudentExamProfile>(`/staff/students/${studentId}/exam-profile/confirm`, {
    method: "POST",
  });
}

export function getStudentExamProfile() {
  return getExamProfileByPath("/student/exam-profile");
}

File: backend/app/schemas/exam_profile.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

EnglishTrack = Literal["english_1", "english_2"]
MathTrack = Literal["math_1", "math_2", "none"]
CetStatus = Literal["not_taken", "cet4", "cet6"]
MathMasteryLevel = Literal["zero", "basic", "good", "strong"]


@dataclass(frozen=True)
class EffectiveExamProfile:
    major_category_code: str
    major_code: str
    major_name: str
    english_track: EnglishTrack
    math_track: MathTrack
    subject_codes: list[str]
    cet_status: CetStatus | None
    cet_score: int | None
    math_mastery_level: MathMasteryLevel | None
    profile_completed_at: datetime | None


class ExamMajorCategoryOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    code: str
    name: str
    sort_order: int


class ExamMajorOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    code: str
    category_code: str
    name: str
    default_english_track: EnglishTrack
    default_math_track: MathTrack
    default_subject_codes: list[str]
    notes: str | None = None


class ExamProfileIn(BaseModel):
    major_category_code: str = Field(min_length=1, max_length=60)
    major_code: str = Field(min_length=1, max_length=60)
    english_track: EnglishTrack | None = None
    math_track: MathTrack | None = None
    subject_codes: list[str] | None = None
    cet_status: CetStatus | None = None
    cet_score: int | None = Field(default=None, ge=0)
    math_mastery_level: MathMasteryLevel | None = None


class ExamProfileOut(BaseModel):
    major_category_code: str
    major_code: str
    major_name: str
    english_track: EnglishTrack | None = None
    math_track: MathTrack | None = None
    effective_english_track: EnglishTrack
    effective_math_track: MathTrack
    subject_codes: list[str]
    cet_status: CetStatus | None = None
    cet_score: int | None = None
    math_mastery_level: MathMasteryLevel | None = None
    profile_completed_at: datetime | None = None
    is_complete: bool
